File: apps/stock-game/app/src/server/testing/fakeProvider.ts
import type { Bar, Interval, Quote, SymbolSearchResult } from '@stock-game/shared'
import type { PriceProvider } from '../providers/types'

function filterBars(bars: Bar[], from: number, to: number): Bar[] {
  return bars.filter((bar) => bar.time >= from && bar.time <= to)
}

export function fakeProvider(overrides?: { quote?: Quote; bars?: Bar[] }): PriceProvider {
  return {
    id: 'fake',
    async getQuote(symbol: string): Promise<Quote> {
      return overrides?.quote ?? { symbol, name: symbol, price: 50, currency: 'USD', exchange: 'TEST', time: 0, delayMinutes: 0 }
    },
    async getBars(_symbol: string, _interval: Interval, from: number, to: number): Promise<Bar[]> {
      return filterBars(overrides?.bars ?? [], from, to)
    },
    async search(_query: string): Promise<SymbolSearchResult[]> {
      return []
    },
  }
}

export function dayBar(date: string, close: number, overrides?: { open?: number; high?: number; low?: number }): Bar {
  const time = Date.parse(date) + 14 * 60 * 60 * 1000 + 30 * 60 * 1000
  const open = overrides?.open ?? close
  const high = overrides?.high ?? close
  const low = overrides?.low ?? close
  return { time, open, high, low, close, volume: 1000 }
}

File: apps/stock-game-api/src/test/kotlin/stockgame/trading/FakeProvider.kt
package stockgame.trading

import stockgame.domain.Bar
import stockgame.domain.Quote
import stockgame.domain.SymbolHit
import stockgame.provider.PriceProvider

class FakeProvider(
    private val bars: List<Bar> = emptyList(),
    private val quote: Quote? = null,
) : PriceProvider {
    override val id: String = "fake"

    override fun getQuote(symbol: String): Quote =
        quote ?: Quote(symbol, symbol, 50.0, "USD", "TEST", 0, 0)

    override fun getBars(symbol: String, interval: String, from: Long, to: Long): List<Bar> =
        bars.filter { it.time in from..to }

    override fun search(query: String): List<SymbolHit> = emptyList()
}

fun dayBar(date: String, close: Double, low: Double = close, high: Double = close): Bar {
    val time = java.time.Instant.parse("${date}T14:30:00Z").toEpochMilli()
    return Bar(time, close, high, low, close, 1000)
}
